telemetry/data_feature_usage.go

```diff
@@ -0,0 +1,242 @@
+			fallthrough
+		case ratio >= 1:
+			CoprocessorCacheTelemetry.MinuteWindow[length-1].P100.Add(1)
+		}
+	} else {
+		CoprocessorCacheTelemetry.MinuteWindow[length-1].P0.Add(1)
+	}
+	CoprocessorCacheTelemetry.Lock.Unlock()
+
+	TiFlashUsageTelemetry.Lock.Lock()
+	length = len(TiFlashUsageTelemetry.MinuteWindow)
+	if length == 0 || time.Since(*TiFlashUsageTelemetry.MinuteWindow[length-1].BeginAt) >= time.Minute {
+		var i int
+		for i = 0; i < length && time.Since(*TiFlashUsageTelemetry.MinuteWindow[i].BeginAt) >= ReportInterval; i++ {
+		}
+		TiFlashUsageTelemetry.MinuteWindow = TiFlashUsageTelemetry.MinuteWindow[i:]
+		TiFlashUsageTelemetry.MinuteWindow = append(TiFlashUsageTelemetry.MinuteWindow, TiFlashUsageItem{})
+		length -= i - 1
+		TiFlashUsageTelemetry.MinuteWindow[length-1].BeginAt = new(time.Time)
+		*TiFlashUsageTelemetry.MinuteWindow[length-1].BeginAt = time.Now()
+	}
+	TiFlashUsageTelemetry.MinuteWindow[length-1].TotalNumbers.Add(1)
+	if task.TiFlashPushDown {
+		TiFlashUsageTelemetry.MinuteWindow[length-1].TiFlashPushDownNumbers.Add(1)
+	}
+	if task.TiFLashExchangePushDown {
+		TiFlashUsageTelemetry.MinuteWindow[length-1].TiFlashExchangePushDownNumbers.Add(1)
+	}
+	TiFlashUsageTelemetry.Lock.Unlock()
+}
+
+// CoprocessorCacheTelemetry is to save the global coprocessor cache telemetry data.
+var CoprocessorCacheTelemetry = struct {
+	MinuteWindow []CoprCacheUsedWindowItem
+	Lock         sync.RWMutex
+}{Lock: sync.RWMutex{}}
+
+// CoprCacheUsedWindowItem is the coprocessor cache telemetry data struct.
+type CoprCacheUsedWindowItem struct {
+	P0   atomic.Uint64 `json:"gte0"`
+	P1   atomic.Uint64 `json:"gte1"`
+	P10  atomic.Uint64 `json:"gte10"`
+	P20  atomic.Uint64 `json:"gte20"`
+	P40  atomic.Uint64 `json:"gte40"`
+	P80  atomic.Uint64 `json:"gte80"`
+	P100 atomic.Uint64 `json:"gte100"`
+
+	BeginAt *time.Time `json:"beginAt"`
+}
+
+// TiFlashUsageTelemetry is to save the global TiFlash usage telemetry data.
+var TiFlashUsageTelemetry = struct {
+	MinuteWindow []TiFlashUsageItem
+	Lock         sync.RWMutex
+}{Lock: sync.RWMutex{}}
+
+// TiFlashUsageItem is the TiFlash usage telemetry data struct.
+type TiFlashUsageItem struct {
+	TotalNumbers                   atomic.Uint64 `json:"totalNumbers"`
+	TiFlashPushDownNumbers         atomic.Uint64 `json:"tiFlashPushDownNumbers"`
+	TiFlashExchangePushDownNumbers atomic.Uint64 `json:"tiFlashExchangePushDownNumbers"`
+
+	BeginAt *time.Time `json:"beginAt"`
+}
+
+func getTelemetryFeatureUsageInfo(ctx sessionctx.Context) (*featureUsageInfo, error) {
+	// init
+	usageInfo := featureUsageInfo{
+		CoprCacheUsed:    make([]*CoprCacheUsedWindowItem, 0, ReportInterval/time.Hour),
+		ClusterIndexUsed: make(map[string]bool),
+		TiFlashUsed:      make([]*TiFlashUsageItem, 0, ReportInterval/time.Hour),
+	}
+
+	// coprocessor cache
+	CoprocessorCacheTelemetry.Lock.Lock()
+	maxLen := 0
+	for _, window := range CoprocessorCacheTelemetry.MinuteWindow {
```

Comment:
Is it hourWindow instead?

Reply:
If window width is `time.Hour`, when we report telemetry data. It may miss 59 minutes of data.